gridscale/resource_gridscale_paas.go

```diff
@@ -0,0 +1,335 @@
+	d.Set("name", props.Name)
+	if creds != nil && len(creds) > 0 {
+		d.Set("username", creds[0].Username)
+		d.Set("password", creds[0].Password)
+	}
+	d.Set("security_zone_uuid", props.SecurityZoneUUID)
+	d.Set("service_template_uuid", props.ServiceTemplateUUID)
+	d.Set("usage_in_minute", props.UsageInMinutes)
+	d.Set("current_price", props.CurrentPrice)
+	d.Set("change_time", props.ChangeTime)
+	d.Set("create_time", props.CreateTime)
+	d.Set("status", props.Status)
+
+	//Get listen ports
+	listenPorts := make([]interface{}, 0)
+	for _, value := range props.ListenPorts {
+		for k, portValue := range value {
+			port := map[string]interface{}{
+				"name": k,
+				"port": portValue,
+			}
+			listenPorts = append(listenPorts, port)
+		}
+	}
+	if err = d.Set("listen_port", listenPorts); err != nil {
+		return fmt.Errorf("Error setting listen ports: %v", err)
+	}
+
+	//Get parameters
+	parameters := make([]interface{}, 0)
+	for k, value := range props.Parameters {
+		param := map[string]interface{}{
+			"param": k,
+			"value": value,
+		}
+		parameters = append(parameters, param)
+	}
+	if err = d.Set("parameter", parameters); err != nil {
+		return fmt.Errorf("Error setting parameters: %v", err)
+	}
+
+	//Get resource limits
+	resourceLimits := make([]interface{}, 0)
+	for _, value := range props.ResourceLimits {
+		limit := map[string]interface{}{
+			"resource": value.Resource,
+			"limit":    value.Limit,
+		}
+		resourceLimits = append(resourceLimits, limit)
+	}
+	if err = d.Set("resource_limit", resourceLimits); err != nil {
+		return fmt.Errorf("Error setting resource limits: %v", err)
+	}
+
+	//Set labels
+	if err = d.Set("labels", props.Labels); err != nil {
+		return fmt.Errorf("Error setting labels: %v", err)
+	}
+
+	//Get all available networks
+	networks, err := client.GetNetworkList(emptyCtx)
+	if err != nil {
+		return fmt.Errorf("Error getting networks: %v", err)
+	}
+	//look for a network that the PaaS service is in
+	for _, network := range networks {
+		securityZones := network.Properties.Relations.PaaSSecurityZones
+		//Each network can contain only one security zone
+		if len(securityZones) >= 1 {
+			if securityZones[0].ObjectUUID == props.SecurityZoneUUID {
+				d.Set("network_uuid", network.Properties.ObjectUUID)
+			}
+		}
+	}
+	return nil
+}
+
+func resourceGridscalePaaSServiceCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gsclient.Client)
+	requestBody := gsclient.PaaSServiceCreateRequest{
+		Name:                    d.Get("name").(string),
+		PaaSServiceTemplateUUID: d.Get("service_template_uuid").(string),
+		Labels:                  convSOStrings(d.Get("labels").(*schema.Set).List()),
+		PaaSSecurityZoneUUID:    d.Get("security_zone_uuid").(string),
+	}
+
+	params := make(map[string]interface{}, 0)
+	for _, value := range d.Get("parameter").(*schema.Set).List() {
+		mapVal := value.(map[string]interface{})
+		var param string
+		var val interface{}
+		for k, v := range mapVal {
```

Comment:
I'm not sure why should go through the loop to look up two values!.
`
if v, found := mapVal["param"]; found {
   param = v.(string)
}
if v, found := mapVal["value"]; found {
   val = v
}
`

Reply:
You are right. I will use `mapVal[key]` directly as they are required in the schema, no need to check if they exist or not.